backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1749 @@
+		fs.Infof(srcDir, "removing empty directory")
+		IDs = append(IDs, srcDir.ID())
+		// API doesn't allow to delete a large number of objects at once, so doing it in chunked
+		if len(IDs) >= api.ListLimit {
+			if err = f.deleteObjects(ctx, IDs, true); err != nil {
+				return err
+			}
+			IDs = nil
+		}
+	}
+	if err := f.deleteObjects(ctx, IDs, true); err != nil {
+		return fmt.Errorf("MergeDirs failed to rmdir: %w", err)
+	}
+	return nil
+}
+
+func (f *Fs) uploadByForm(ctx context.Context, in io.Reader, name string, size int64, form *api.Form, options ...fs.OpenOption) (err error) {
+	// struct to map. transferring values from MultParts to url parameter
+	params := url.Values{}
+	iVal := reflect.ValueOf(&form.MultiParts).Elem()
+	iTyp := iVal.Type()
+	for i := 0; i < iVal.NumField(); i++ {
+		params.Set(iTyp.Field(i).Tag.Get("json"), iVal.Field(i).String())
+	}
+	formReader, contentType, overhead, err := rest.MultipartUpload(ctx, in, params, "file", name)
+	if err != nil {
+		return fmt.Errorf("failed to make multipart upload: %w", err)
+	}
+
+	contentLength := overhead + size
+	opts := rest.Opts{
+		Method:           form.Method,
+		RootURL:          form.URL,
+		Body:             formReader,
+		ContentType:      contentType,
+		ContentLength:    &contentLength,
+		Options:          options,
+		TransferEncoding: []string{"identity"},
+		NoResponse:       true,
+	}
+
+	var resp *http.Response
+	err = f.pacer.CallNoRetry(func() (bool, error) {
+		resp, err = f.srv.Call(ctx, &opts)
+		return f.shouldRetry(ctx, resp, err)
+	})
+	return
+}
+
+func (f *Fs) uploadByResumable(ctx context.Context, in io.Reader, resumable *api.Resumable, options ...fs.OpenOption) (err error) {
+	p := resumable.Params
+	endpoint := strings.Join(strings.Split(p.Endpoint, ".")[1:], ".") // "mypikpak.com"
+
+	cfg := &aws.Config{
+		Credentials: credentials.NewStaticCredentials(p.AccessKeyID, p.AccessKeySecret, p.SecurityToken),
+		Region:      aws.String("pikpak"),
+		Endpoint:    &endpoint,
+	}
+	sess, err := session.NewSession(cfg)
+	if err != nil {
+		return
+	}
+
+	uploader := s3manager.NewUploader(sess)
+	// Upload input parameters
+	uParams := &s3manager.UploadInput{
+		Bucket: &p.Bucket,
+		Key:    &p.Key,
+		Body:   in,
+	}
+	// Perform upload with options different than the those in the Uploader.
+	_, err = uploader.UploadWithContext(ctx, uParams, func(u *s3manager.Uploader) {
+		// TODO can be user-configurable
+		u.PartSize = 10 * 1024 * 1024 // 10MB part size
+	})
+	return
+}
+
+func (f *Fs) upload(ctx context.Context, in io.Reader, leaf, dirID string, size int64, options ...fs.OpenOption) (*api.File, error) {
+	// Calculate sha1sum; grabbed from package jottacloud
+	//
+	// unwrap the accounting from the input, we use wrap to put it
+	// back on after the buffering
+	var wrap accounting.WrapFn
+	in, wrap = accounting.UnWrap(in)
+	var cleanup func()
+	sha1Str, in, cleanup, err := readSHA1(in, size, int64(f.opt.HashMemoryThreshold))
```

Comment:
Note that some source objects (eg files on the local disk) will have an SHA1 available so you don't need to read it.

You should try to read it first like the jottacloud code does

https://github.com/rclone/rclone/blob/9baa4d1c3c5b592510af75495c718faadeaf9cd0/backend/jottacloud/jottacloud.go#L1853-L1867